clang/lib/Sema/SemaExpr.cpp

```diff
@@ -9966,8 +9966,13 @@ AssignConvertType Sema::CheckSingleAssignmentConstraints(QualType LHSType,
       // If there is a conversion of some kind, check to see what kind of
       // pointer conversion happened so we can diagnose a C++ compatibility
       // diagnostic if the conversion is invalid. This only matters if the RHS
-      // is some kind of void pointer.
-      if (Kind != CK_NoOp && !getLangOpts().CPlusPlus) {
+      // is some kind of void pointer. We have a carve-out when the RHS is from
+      // a macro expansion because the use of a macro may indicate different
+      // code between C and C++. Consider: char *s = NULL; where NULL is
+      // defined as (void *)0 in C (which would be invalid in C++), but 0 in
+      // C++, which is valid in C++.
+      if (Kind != CK_NoOp && !getLangOpts().CPlusPlus &&
+          !RHS.get()->getBeginLoc().isMacroID()) {
```

Comment:
This is pretty liberal, right?  Why EVERYTHING? 

Why not just use the common patterns of NULL that we care about?  There is a function somewhere to test that it is the 'C" style null-pointer-cast (`Expr::isNullPointerConstant`) that does exactly what you want I think.

Reply:
MOST of those would work with the `isNullPointerConstant` I would think, we get the `(void*)0` pattern with that, no?    If not, it absolutely should...

Reply:
That's kind of the point -- this code is specific to the case of a void pointer conversion from a null pointer constant.

So we do get the pattern, pass that predicate and get to the point of saying "cool, time to diagnose this implicit void * cast" and now we're saying "oh, but the RHS is a macro so nevermind, just be quiet about it because the macro may expand differently in C++ than it did in C".

Reply:
OH! I  misunderstood what was happening here/missed the call to `isNullPointerConstant` above.  I thought this was forgiving ALL `char* F = MY_MAGIC_MACRO`'s.

Reply:
Ah, nope! We've got a test for that:
```
 #define SOMETHING_THAT_IS_NOT_NULL (void *)12

  char *ptr3 = SOMETHING_THAT_IS_NOT_NULL; // c-warning {{implicit conversion when initializing 'char *' with an expression of type 'void *' is not permitted in C++}} \
                                              cxx-error {{cannot initialize a variable of type 'char *' with an rvalue of type 'void *'}}

  ...

  ptr3 = SOMETHING_THAT_IS_NOT_NULL; // c-warning {{implicit conversion when assigning to 'char *' from type 'void *' is not permitted in C++}} \
                                        cxx-error {{assigning to 'char *' from incompatible type 'void *'}}
```